### PR TITLE
Implement remaining pieces for LunarG API-Samples

### PR DESCRIPTION
**Pull request by msiglreith**
Requires the queue present PR to be merged and rebasing of the portability branch before merging!

### CODE REVIEW

File: libportability-gfx/src/conv.rs [hunk 1]
```diff
@@ -430,3 +431,95 @@ pub fn map_image_acces(access: VkAccessFlags) -> image::Access {
 
     mask
 }
+
+pub fn map_subpass_contents(contents: VkSubpassContents) -> command::SubpassContents {
+    match contents {
+        VkSubpassContents::VK_SUBPASS_CONTENTS_INLINE => command::SubpassContents::Inline,
+        VkSubpassContents::VK_SUBPASS_CONTENTS_SECONDARY_COMMAND_BUFFERS =>
+            command::SubpassContents::SecondaryBuffers,
+
```
Comment (kvark) [inline libportability-gfx/src/conv.rs]: nit: weird newline

File: libportability-gfx/src/conv.rs [hunk 2]
```diff
@@ -430,3 +431,95 @@ pub fn map_image_acces(access: VkAccessFlags) -> image::Access {
...
+        VkPolygonMode::VK_POLYGON_MODE_LINE => pso::PolygonMode::Line(line_width),
+        VkPolygonMode::VK_POLYGON_MODE_POINT => pso::PolygonMode::Point,
+        _ => panic!("Unexpected polygon mode: {:?}", mode),
+    }
+}
+
+pub fn map_cull_face(cull: VkCullModeFlags) -> Option<pso::CullFace> {
+    use super::VkCullModeFlagBits::*;
+
+    if cull == VK_CULL_MODE_NONE as _ { None }
```
Comment (kvark) [inline libportability-gfx/src/conv.rs]: let's have a `match` here

File: libportability-gfx/src/conv.rs [hunk 3]
```diff
@@ -430,3 +431,95 @@ pub fn map_image_acces(access: VkAccessFlags) -> image::Access {
...
+    else if cull == VK_CULL_MODE_FRONT_BIT as _ { Some(pso::CullFace::Front) }
+    else if cull == VK_CULL_MODE_BACK_BIT as _ { Some(pso::CullFace::Back) }
+    else if cull == VK_CULL_MODE_FRONT_AND_BACK as _ { unimplemented!() } // TODO: can we support it?
```
Comment (kvark) [inline libportability-gfx/src/conv.rs]: we can, needs the cull face to become flags then

File: libportability-gfx/src/impls.rs [hunk 4]
```diff
@@ -949,14 +952,345 @@ pub extern "C" fn gfxMergePipelineCaches(
...
+                    let data = slice::from_raw_parts(
+                        specialization.pData,
+                        specialization.dataSize as _,
+                    );
+                    let entries = slice::from_raw_parts(
+                        specialization.pMapEntries,
+                        specialization.mapEntryCount as _,
+                    );
+
+                    entries.
```
Comment (kvark) [inline libportability-gfx/src/impls.rs]: nit: move the '.' to the new line

File: libportability-gfx/src/impls.rs [hunk 5]
```diff
@@ -949,14 +952,345 @@ pub extern "C" fn gfxMergePipelineCaches(
...
+
+            for stage in stages {
+                use super::VkShaderStageFlagBits::*;
+
+                let (ref name, ref specialization) = shader_stages[cur_shader_stage];
+                cur_shader_stage += 1;
+
+                let entry_point = pso::EntryPoint {
+                    entry: &name,
+                    module: stage.module.deref(),
```
Comment (kvark) [inline libportability-gfx/src/impls.rs]: note: you can achieve the same without importing `Deref` by doing `&*stage.module`

File: libportability-gfx/src/impls.rs [hunk 6]
```diff
@@ -949,14 +952,345 @@ pub extern "C" fn gfxMergePipelineCaches(
...
+                    }
+                })
+                .collect::<Vec<_>>();
+
+            (bindings, attributes)
+        };
+
+        let input_assembler = {
+            let input_state = unsafe { &*info.pInputAssemblyState };
+            let tessellation_state = if shaders.hull.is_some() {
```
Comment (kvark) [inline libportability-gfx/src/impls.rs]: use `map()`?

File: libportability-gfx/src/impls.rs [hunk 7]
```diff
@@ -949,14 +952,345 @@ pub extern "C" fn gfxMergePipelineCaches(
...
+            }
+        };
+
+        // TODO: `pColorBlendState` could contain garbage, but implementations
+        //        can ignore it in some circumstances. How to handle it?
+        let blender = {
+            let mut blend_desc = pso::BlendDesc::default();
+
+            let blend_state = unsafe { info.pColorBlendState.as_ref() };
+            if let Some(state) = blend_state {
```
Comment (kvark) [inline libportability-gfx/src/impls.rs]: nit: could get away without `blend_state`

File: libportability-gfx/src/impls.rs [hunk 8]
```diff
@@ -949,14 +952,345 @@ pub extern "C" fn gfxMergePipelineCaches(
...
+                        } else {
+                            pso::BlendState::Off
+                        };
+
+                        pso::ColorBlendDesc(color_mask, blend)
+                    })
+                    .collect();
+            }
+
+            // TODO: multisampling
```
Comment (kvark) [inline libportability-gfx/src/impls.rs]: add an assert for incoming multisampling?

File: libportability-gfx/src/impls.rs [hunk 9]
```diff
@@ -949,14 +952,345 @@ pub extern "C" fn gfxMergePipelineCaches(
...
+
+            blend_desc
+        };
+
+        // TODO: `pDepthStencilState` could contain garbage, but implementations
+        //        can ignore it in some circumstances. How to handle it?
```
Comment (kvark) [inline libportability-gfx/src/impls.rs]: what are those circumstances/cases?

File: libportability-gfx/src/impls.rs [hunk 10]
```diff
@@ -1695,23 +2094,49 @@ pub extern "C" fn gfxCmdBindIndexBuffer(
...
+    };
+    let offsets = unsafe {
+        slice::from_raw_parts(pOffsets, bindingCount as _)
+    };
+
+    let views = buffers
+        .into_iter()
+        .zip(offsets.into_iter())
+        .map(|(buffer, offset)| {
+            let offset = *offset as _;
```
Comment (kvark) [inline libportability-gfx/src/impls.rs]: FWIW, I'd like us to not use shadowing too much

File: libportability-gfx/src/impls.rs [hunk 11]
```diff
@@ -1695,23 +2094,49 @@ pub extern "C" fn gfxCmdBindIndexBuffer(
...
+            let buffer = match **buffer {
+                Buffer::Buffer(ref buffer) => buffer,
+                Buffer::Unbound(_) => panic!("Non-sparse buffers need to be bound to device memory."),
+            };
+
+            (buffer, offset)
+        })
+        .collect::<Vec<_>>();
```
Comment (kvark) [inline libportability-gfx/src/impls.rs]: type shouldn't be needed here

File: libportability-gfx/src/conv.rs (hunk 2 above)
Comment (msiglreith) [reply libportability-gfx/src/conv.rs]: It's not that easy as types mismatch iirc.

File: libportability-gfx/src/conv.rs (hunk 3 above)
Comment (msiglreith) [reply libportability-gfx/src/conv.rs]: DX12 for example doesn't support it as far as I know.

File: libportability-gfx/src/impls.rs (hunk 5 above)
Comment (msiglreith) [reply libportability-gfx/src/impls.rs]: I think it was `&**stage` for these deref cases, not sure what's nicer..

File: libportability-gfx/src/impls.rs (hunk 9 above)
Comment (msiglreith) [reply libportability-gfx/src/impls.rs]: If the subpass doesn't use a depth-stencil attachment or rasterization is disabled (this part we can check) `pDepthStencilState` is ignored. The specification doesn't explicitly enforce that it must be NULL.

File: libportability-gfx/src/conv.rs (hunk 2 above)
Comment (kvark) [reply libportability-gfx/src/conv.rs]: You can do `match VkCullModeFlagBits::from_bits(cull) {..}`

File: libportability-gfx/src/conv.rs (hunk 3 above)
Comment (kvark) [reply libportability-gfx/src/conv.rs]: yes, but we are Vulkan portability :)
on DX12, we'll treat it as disabled rasterizer

File: libportability-gfx/src/impls.rs (hunk 5 above)
Comment (kvark) [reply libportability-gfx/src/impls.rs]: Arguably, this is still more explicit than `x.deref()` :)

File: libportability-gfx/src/conv.rs (hunk 3 above)
Comment (msiglreith) [reply libportability-gfx/src/conv.rs]: hm, yes, that would emulate the behavior as we have the topological information about the primitives and can discard primitives if we face this cull mode and use triangles.
The only issue I currently see are pipeline statistic queries, as primitive discard happens in an earlier state than clipping (discard -> .. -> clipping -> .. -> culling), affecting the number of primitives reaching the clipping state in contrast to pure back&front culling.

File: libportability-gfx/src/conv.rs (hunk 2 above)
Comment (msiglreith) [reply libportability-gfx/src/conv.rs]: wait, we don't use `bitflags` :D 
well, I could `transmute` + truncate it?

**Comment by msiglreith** [comment]
bors r=kvark

**Pull request merged**